internal/debug/debug.go

```diff
@@ -0,0 +1,165 @@
+// and returns the DebugPod object. Use the Clean() method to delete it.
+func New(node string) (*DebugPod, error) {
+	errCh := make(chan error)
+	go func() {
+		errCh <- start(node)
+	}()
+	for {
+		select {
+		case err := <-errCh:
+			return nil, err
+		default:
+			return waitForPod()
+		}
+	}
+}
+
+func (dp *DebugPod) Exec(cmd string) ([]byte, error) {
+	cmdOnDebugPod := append([]string{"exec", "-n", dp.Namespace, dp.Name, "--", "chroot", "/host"}, strings.Split(cmd, " ")...)
+	out, err := exec.Command("oc", cmdOnDebugPod...).CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("failed to exec command \"%s\" on node %s: %v\n%s", cmd, dp.NodeName, err, string(out))
+	}
+
+	return out, nil
+}
+
+func (dp *DebugPod) ExecWithRetry(cmd string) ([]byte, error) {
+	var (
+		out     []byte
+		execErr error
+	)
+
+	if err := wait.PollUntilContextTimeout(context.TODO(), interval, duration, true, func(ctx context.Context) (bool, error) {
+		out, execErr = dp.Exec(cmd)
+		if execErr != nil {
+			return false, execErr
+		}
+
+		return true, execErr
+	}); err != nil {
+		return nil, errors.Join(err, execErr)
+	}
+
+	return out, nil
+}
+
+func (dp *DebugPod) Clean() error {
+	output, err := exec.Command("oc", "delete", "pod", "-n", dp.Namespace, dp.Name).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed deleting debug pod %s/%s: %v\n%s", dp.Namespace, dp.Name, err, string(output))
+	}
+
+	output, err = exec.Command("oc", "delete", "ns", dp.Namespace).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed deleting debug namespace %s: %v\n%s", dp.Namespace, err, string(output))
+	}
+
+	return nil
+}
+
+func (dp *DebugPod) String() string {
+	return fmt.Sprintf(dp.Name)
+}
+
+// start runs the `oc debug` command on a given node and enters sleep infinity
+func start(node string) error {
+	cmd := []string{"debug", fmt.Sprintf("node/%s", node), "--", "/bin/bash", "-c", "trap : TERM INT; sleep infinity & wait"}
+	output, err := exec.Command("oc", cmd...).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("command \"oc %s\" returned error: %s\n%s", strings.Join(cmd, " "), err, string(output))
+	}
+	return nil
+}
+
+func waitForPod() (*DebugPod, error) {
+	var (
+		pod    *DebugPod
+		getErr error
+	)
+
+	err := wait.PollUntilContextTimeout(context.TODO(), interval, time.Second*30, true, func(ctx context.Context) (bool, error) {
+		pod, getErr = get()
+		if getErr != nil && strings.Contains(getErr.Error(), "not found") {
+			return false, nil // pod not ready, retry.
+		}
+		if getErr != nil && strings.Contains(getErr.Error(), "not running") {
+			return false, nil // pod not ready, retry.
+		}
+		if getErr != nil {
+			return true, getErr // valid error, stop trying.
+		}
+		return true, nil
+	})
+	if err != nil {
+		return nil, errors.Join(err, getErr)
+	}
+
+	return pod, nil
+}
+
+// get returns the debug pod that is present in the cluster
+func get() (*DebugPod, error) {
```

Comment:
I think this function could be simplified by using a client and not oc

Reply:
Yep, I agree